R/spark_compile.R

```diff
@@ -97,29 +81,174 @@ spark_compile <- function(name, spark_home) {
 
   # ensure 'inst/java' exists
   inst_java_path <- file.path(root, "inst/java")
-  if (!file.exists(inst_java_path))
-    if (!dir.create(inst_java_path, recursive = TRUE))
-      stop("failed to create directory '", inst_java_path, "'")
+  ensure_directory(inst_java_path)
 
-  # call 'scalac' compiler
+  # call 'scalac' with CLASSPATH set
   classpath <- Sys.getenv("CLASSPATH")
-
-  # set CLASSPATH environment variable rather than passing
-  # in on command line (mostly aesthetic)
   Sys.setenv(CLASSPATH = CLASSPATH)
-  execute("scalac", paste(shQuote(scala_files), collapse = " "))
-  Sys.setenv(CLASSPATH = classpath)
+  on.exit(Sys.setenv(CLASSPATH = classpath), add = TRUE)
+  status <- execute("scalac", paste(shQuote(scala_files), collapse = " "))
+  if (status)
+    stop("==> failed to compile Scala source files")
 
   # call 'jar' to create our jar
-  class_files <- file.path(name, list.files(name, pattern = "class$"))
-  execute("jar cf", jar_path, paste(shQuote(class_files), collapse = " "))
+  class_files <- list.files(pattern = "class$", recursive = TRUE, full.names = TRUE)
+  class_files <- sub("./", "", class_files)
+  status <- execute("jar cf", jar_path, paste(shQuote(class_files), collapse = " "))
+  if (status)
+    stop("==> failed to build Java Archive")
 
   # double-check existence of jar
-  if (file.exists(jar_path)) {
-    message("*** ", basename(jar_path), " successfully created.")
-  } else {
-    stop("*** failed to create ", jar_name)
+  if (!file.exists(jar_path))
+    stop("==> failed to create ", jar_name)
+
+  message("==> ", basename(jar_path), " successfully created\n")
+  TRUE
+}
+
+#' Compile Scala sources into a Java Archive (jar)
+#'
+#' Compile the \code{scala} source files contained within an \R package
+#' into a Java Archive (\code{jar}) file that can be loaded and used within
+#' a Spark environment.
+#'
+#' @param package The path to an \R package.
+#' @param spark_versions The Spark versions to build against. When \code{NULL},
+#'   builds against all Spark versions discovered with
+#'   \code{\link{spark_available_versions}}.
+#' @param scalac The path to the \code{scalac} compiler to be used. When unset,
+#'   \code{scalac} will be discovered on the PATH.
+#' @param jar The path to the \code{jar} Java archive tool to be used.
+#'   When unset, \code{jar} will be discovered on the PATH.
+#'
+#' @import rprojroot
+#' @import digest
+#'
+#' @keywords internal
+#' @export
+compile_package_jars <- function(package = rprojroot::find_package_root_file(),
+                                 compilation_spec = default_compilation_spec())
```

Comment:
If we don't want to export `default_compilation_spec` we might just consider letting this be `"default"` and calling that function internally.